debezium-server-milvus/src/main/java/io/debezium/server/milvus/MilvusChangeConsumer.java

```diff
@@ -0,0 +1,207 @@
+
+            final var sourceRecord = toSourceRecord(record);
+
+            // Milvus does not support dots in collection names so we by default replace them
+            // with underscores so the user does not need to provide router or mapper
+            final var collectionName = streamNameMapper.map(record.destination()).replace('.', '_');
+
+            if (isSchemaChange(sourceRecord)) {
+                LOGGER.debug("Schema change event, ignoring it");
+                committer.markProcessed(record);
+                continue;
+            }
+
+            if (record.key() == null) {
+                throw new DebeziumException("Milvus does not support collections without primary key");
+            }
+            schema.validateKey(collectionName, sourceRecord.keySchema());
+
+            if (sourceRecord.value() == null) {
+                deleteRecord(collectionName, record, committer);
+            }
+            else if (Envelope.isEnvelopeSchema(sourceRecord.valueSchema())) {
+                final var valueStruct = (Struct) sourceRecord.value();
+                switch (Operation.forCode(valueStruct.getString(Envelope.FieldName.OPERATION))) {
+                    case Operation.READ:
+                    case Operation.CREATE:
+                    case Operation.UPDATE:
+                        upsertRecord(collectionName, record, committer);
+                        break;
+                    case Operation.DELETE:
+                        deleteRecord(collectionName, record, committer);
+                        break;
+                    default:
+                        LOGGER.info("Unsupported operation, skipping record '{}'", record);
+                }
+            }
+            else {
+                // Extracted new record state
+                upsertRecord(collectionName, record, committer);
+            }
+        }
+
+        committer.markBatchFinished();
+    }
+
+    protected SourceRecord toSourceRecord(final ChangeEvent<Object, Object> record) {
+        // Milvus sink requires access to the message schema so it can do schema evolution
+        // This is not a part of public Debezium Engine API but is an implementation detail on
+        // which Debezium Server can rely
+        @SuppressWarnings("rawtypes")
+        final var sourceRecord = ((EmbeddedEngineChangeEvent) record).sourceRecord();
+        return sourceRecord;
+    }
+
+    private void upsertRecord(String collectionName, ChangeEvent<Object, Object> record,
+                              RecordCommitter<ChangeEvent<Object, Object>> committer)
+            throws InterruptedException {
+        final var data = getValue(record, toSourceRecord(record));
+
+        final var request = UpsertReq.builder()
+                .collectionName(collectionName)
+                .data(List.of(data))
+                .build();
+        milvusClient.upsert(request);
+        committer.markProcessed(record);
```

Comment:
shouldn't we check `upsert()` result before marking record as processed (same for delete)?

Reply:
Good point, when I was doing some testing during RnD phase I was getting exceptions for which this is safe approach but definitely better to check the response too.

Reply:
Well, I checked the API and it seems there is no status retunred, only the number of records modified. In this case I think we can ignore the results as zero result can be valid in some situations too.

Reply:
I thought it [returns status](https://github.com/milvus-io/milvus-sdk-java/blob/master/sdk-core/src/main/java/io/milvus/client/MilvusClient.java#L414), but I didn't check it, so yes, in such case this is completely fine.